File: isolation/artifacts/hexhunter/hexhunter.js
/* hexhunter — JavaScript port of libhexhunter.
 *
 * Pure JS, no dependencies.  Runs in browsers and Node.  Byte-for-byte
 * compatible with the C library (`hexhunter.c`) and Python port
 * (`hexhunter.py`) for the same (population, generations, rng_seed)
 * configuration.
 *
 * Designed to live on a page or in a Web Worker — the GA loop yields
 * to the event loop after every generation so the UI stays responsive.
 *
 * Public API:
 *
 *     // synchronous (returns Uint8Array of length 4096)
 *     const out = hexhunter({population: 30, generations: 40, rng_seed: 42});
 *     const out = hexhunter();                  // all defaults
 *     const r   = hexhunter_refine(out, {generations: 20});
 *     const f   = hexhunter_fitness(r);
 *
 *     // async (yields after each generation; takes onProgress callback)
 *     const out = await hexhunter_async({population: 30, generations: 40,
 *                                          progress: (g, total, best, mean, tail) => ... });
 */

(function (root, factory) {
    if (typeof module === 'object' && module.exports) module.exports = factory();
    else root.hexhunter_lib = factory();
}(typeof self !== 'undefined' ? self : this, function () {
'use strict';

const K              = 4;
const NSIT           = 16384;
const GENOME_BYTES   = 4096;

const DEF_POP                = 30;
const DEF_GENS               = 40;
const DEF_INIT_MUT_RATE      = 0.05;
const DEF_BREED_MUT_RATE     = 0.005;
const DEF_GRID_W             = 14;
const DEF_GRID_H             = 14;
const DEF_HORIZON            = 25;
const DEF_RNG_SEED           = 42;

const DY  = [-1, -1,  0,  0,  1,  1];
const DXE = [ 0,  1, -1,  1, -1,  0];
const DXO = [-1,  0, -1,  1,  0,  1];

/* ── Park-Miller LCG, 32-bit unsigned ─────────────────────────────── */
/* Math.imul + >>>0 give us deterministic wrap-around matching the C
 * `state * 1103515245u + 12345u` semantics.  Verified bit-for-bit
 * against hexhunter.c's hh_rng. */

function rng_make(seed) {
    return { state: ((seed | 0) === 0 ? 1 : (seed >>> 0)) };
}
function rng_u32(r) {
    r.state = (Math.imul(r.state, 1103515245) + 12345) >>> 0;
    return r.state >>> 16;
}
function rng_rand(r) {        // 0..0xFFFF, matches HH_RAND_MAX
    return rng_u32(r) & 0xFFFF;
}
function rng_unit(r) {
    return rng_rand(r) / 0xFFFF;
}

/* ── Packed-genome helpers (2 bits per situation) ─────────────────── */

function g_get(g, idx) {
    return (g[idx >> 2] >> ((idx & 3) * 2)) & 3;
}
function g_set(g, idx, v) {
    const b = idx >> 2, o = (idx & 3) * 2;
    g[b] = (g[b] & ~(3 << o)) | ((v & 3) << o);
}
function sit_idx(self_, n) {
    let i = self_;
    for (let k = 0; k < 6; k++) i = i * K + n[k];
    return i;
}

/* ── Hex stepping (pointy-top, row-parity-sensitive) ─────────────── */

function step_grid(g, src, dst, gw, gh) {
    const n = [0, 0, 0, 0, 0, 0];
    for (let y = 0; y < gh; y++) {
        const dx = (y & 1) ? DXO : DXE;
        for (let x = 0; x < gw; x++) {
            const self_ = src[y * gw + x];
            for (let k = 0; k < 6; k++) {
                const yy = y + DY[k];
                const xx = x + dx[k];
                n[k] = (yy >= 0 && yy < gh && xx >= 0 && xx < gw)
                     ? src[yy * gw + xx] : 0;
            }
            dst[y * gw + x] = g_get(g, sit_idx(self_, n));
        }
    }
}

function seed_grid(grid, gw, gh, seed) {
    const r = rng_make(seed);
    const n = gw * gh;
    for (let i = 0; i < n; i++) grid[i] = rng_u32(r) & 3;
}

/* ── Class-4 fitness (smooth tent peaking at activity=0.12) ──────── */

function fitness_inner(genome, gridSeed, gw, gh, horizon) {
    const n = gw * gh;
    const a = new Uint8Array(n);
    const b = new Uint8Array(n);
    seed_grid(a, gw, gh, gridSeed);
    const act = new Float64Array(horizon);

    for (let t = 0; t < horizon; t++) {
        step_grid(genome, a, b, gw, gh);
        let changed = 0;
        for (let i = 0; i < n; i++) if (a[i] !== b[i]) changed++;
        act[t] = changed / n;
        a.set(b);
    }

    let uniform = true;
    for (let i = 1; i < n; i++) if (a[i] !== a[0]) { uniform = false; break; }
    const counts = [0, 0, 0, 0];
    for (let i = 0; i < n; i++) counts[a[i]]++;
    let diversity = 0;
    for (let c = 0; c < K; c++) if (counts[c] * 100 >= n) diversity++;

    let tailN = (horizon / 3) | 0;
    if (tailN < 1) tailN = 1;
    let avg = 0;
    for (let i = horizon - tailN; i < horizon; i++) avg += act[i];
    avg /= tailN;

    let score = 0;
    if (!uniform) score += 1.0;
    let aperiodic = false;
    for (let i = horizon - tailN; i < horizon; i++) {
        if (act[i] > 0.001) { aperiodic = true; break; }
    }
    if (aperiodic) score += 1.5;

    let activityReward;
    if (avg <= 0.12) activityReward = avg / 0.12;
    else             activityReward = (0.75 - avg) / 0.63;
    if (activityReward < 0) activityReward = 0;
    score += 2.0 * activityReward;

    if (diversity >= 2) {
        score += 0.25 * Math.min(diversity, K);
    }
    return { score: score, activity_tail: avg };
}

/* ── GA ops (use the threaded RNG, not Math.random) ──────────────── */

function mutate(dst, src, rate, r) {
    dst.set(src);
    for (let i = 0; i < NSIT; i++) {
        if (rng_unit(r) < rate) g_set(dst, i, rng_rand(r) & 3);
    }
}
function cross(dst, a, b, r) {
    const cut = 1 + (rng_rand(r) % (GENOME_BYTES - 1));
    dst.set(a.subarray(0, cut));
    dst.set(b.subarray(cut), cut);
}

function identity_genome() {
    const g = new Uint8Array(GENOME_BYTES);
    g.fill(0x00, 0,    1024);
    g.fill(0x55, 1024, 2048);
    g.fill(0xAA, 2048, 3072);
    g.fill(0xFF, 3072, 4096);
    return g;
}

/* ── Config resolution (unset/zero → default) ────────────────────── */

function resolve(cfg) {
    cfg = cfg || {};
    const c = {};
    c.population          = cfg.population          > 0    ? cfg.population          : DEF_POP;
    c.generations         = cfg.generations         > 0    ? cfg.generations         : DEF_GENS;
    c.init_mutation_rate  = cfg.init_mutation_rate  > 0    ? cfg.init_mutation_rate  : DEF_INIT_MUT_RATE;
    c.breed_mutation_rate = cfg.breed_mutation_rate > 0    ? cfg.breed_mutation_rate : DEF_BREED_MUT_RATE;
    c.grid_w              = cfg.grid_w              > 0    ? cfg.grid_w              : DEF_GRID_W;
    c.grid_h              = cfg.grid_h              > 0    ? cfg.grid_h              : DEF_GRID_H;
    c.horizon             = cfg.horizon             > 0    ? cfg.horizon             : DEF_HORIZON;
    c.rng_seed            = cfg.rng_seed            != null ? (cfg.rng_seed >>> 0)   : DEF_RNG_SEED;
    c.progress            = cfg.progress || null;
    return c;
}

/* ── Core GA ─────────────────────────────────────────────────────── */

function score_and_sort(pool, fits, cfg) {
    for (let i = 0; i < pool.length; i++) {
        fits[i] = fitness_inner(pool[i], cfg.rng_seed,
                                 cfg.grid_w, cfg.grid_h, cfg.horizon).score;
    }
    /* Insertion sort by fitness descending. */
    for (let i = 1; i < pool.length; i++) {
        const fv = fits[i];
        const tmp = new Uint8Array(pool[i]);   // copy
        let j = i - 1;
        while (j >= 0 && fits[j] < fv) {
            fits[j + 1] = fits[j];
            pool[j + 1] = new Uint8Array(pool[j]);
            j--;
        }
        fits[j + 1] = fv;
        pool[j + 1] = tmp;
    }
}

function do_breed(pool, cfg, rng) {
    const half = pool.length >> 1;
    for (let i = half; i < pool.length; i++) {
        const pa = rng_rand(rng) % half;
        const pb = rng_rand(rng) % half;
        const tmp = new Uint8Array(GENOME_BYTES);
        cross(tmp, pool[pa], pool[pb], rng);
        const child = new Uint8Array(GENOME_BYTES);
        mutate(child, tmp, cfg.breed_mutation_rate, rng);
        pool[i] = child;
    }
}

function init_pool(seedGenome, cfg, rng) {
    const pop = cfg.population;
    const pool = new Array(pop);
    pool[0] = new Uint8Array(seedGenome);
    for (let i = 1; i < pop; i++) {
        const child = new Uint8Array(GENOME_BYTES);
        mutate(child, pool[0], cfg.init_mutation_rate, rng);
        pool[i] = child;
    }
    return pool;
}

function call_progress(cfg, gen, pool, fits) {
    if (!cfg.progress) return;
    let sum = 0;
    for (let i = 0; i < fits.length; i++) sum += fits[i];
    const tail = fitness_inner(pool[0], cfg.rng_seed,
                                 cfg.grid_w, cfg.grid_h, cfg.horizon).activity_tail;
    cfg.progress(gen, cfg.generations, fits[0], sum / fits.length, tail);
}

function run_sync(cfg, seedGenome) {
    const rng = rng_make(cfg.rng_seed);
    const pool = init_pool(seedGenome, cfg, rng);
    const fits = new Float64Array(cfg.population);

    for (let gen = 0; gen < cfg.generations; gen++) {
        score_and_sort(pool, fits, cfg);
        call_progress(cfg, gen + 1, pool, fits);
        do_breed(pool, cfg, rng);
    }
    score_and_sort(pool, fits, cfg);
    return new Uint8Array(pool[0]);
}

async function run_async(cfg, seedGenome) {
    const rng = rng_make(cfg.rng_seed);
    const pool = init_pool(seedGenome, cfg, rng);
    const fits = new Float64Array(cfg.population);

    for (let gen = 0; gen < cfg.generations; gen++) {
        score_and_sort(pool, fits, cfg);
        call_progress(cfg, gen + 1, pool, fits);
        do_breed(pool, cfg, rng);
        /* Yield once per generation so the host event loop runs (UI
         * stays responsive in the browser). */
        await new Promise(r => setTimeout(r, 0));
    }
    score_and_sort(pool, fits, cfg);
    return new Uint8Array(pool[0]);
}

/* ── Public API ──────────────────────────────────────────────────── */

function hexhunter(cfg) {
    return run_sync(resolve(cfg), identity_genome());
}
function hexhunter_refine(inGenome, cfg) {
    if (!(inGenome && inGenome.length === GENOME_BYTES))
        throw new Error('in_genome must be ' + GENOME_BYTES + ' bytes');
    return run_sync(resolve(cfg), new Uint8Array(inGenome));
}
function hexhunter_async(cfg) {
    return run_async(resolve(cfg), identity_genome());
}
function hexhunter_refine_async(inGenome, cfg) {
    if (!(inGenome && inGenome.length === GENOME_BYTES))
        throw new Error('in_genome must be ' + GENOME_BYTES + ' bytes');
    return run_async(resolve(cfg), new Uint8Array(inGenome));
}
function hexhunter_fitness(genome, cfg) {
    if (!(genome && genome.length === GENOME_BYTES))
        throw new Error('genome must be ' + GENOME_BYTES + ' bytes');
    const c = resolve(cfg);
    return fitness_inner(genome, c.rng_seed, c.grid_w, c.grid_h, c.horizon).score;
}

return {
    K: K, NSIT: NSIT, GENOME_BYTES: GENOME_BYTES,
    DEF_POP: DEF_POP, DEF_GENS: DEF_GENS,
    DEF_INIT_MUT_RATE: DEF_INIT_MUT_RATE,
    DEF_BREED_MUT_RATE: DEF_BREED_MUT_RATE,
    DEF_GRID_W: DEF_GRID_W, DEF_GRID_H: DEF_GRID_H,
    DEF_HORIZON: DEF_HORIZON, DEF_RNG_SEED: DEF_RNG_SEED,
    identity_genome: identity_genome,
    hexhunter: hexhunter,
    hexhunter_refine: hexhunter_refine,
    hexhunter_async: hexhunter_async,
    hexhunter_refine_async: hexhunter_refine_async,
    hexhunter_fitness: hexhunter_fitness,
};
}));

File: isolation/artifacts/hexhunter/hexhunter.py
"""Pure-Python port of libhexhunter.

Mirrors the C library byte-for-byte: same Park-Miller LCG, same
2-bit packing, same hex topology, same smooth-tent class-4 fitness,
same insertion-sort + tournament breeding.  A run with the same
config + rng_seed produces the *same* 4096-byte ruleset as the C
implementation (verified with `verify_against_c.py`).

Slow compared to C — pure-Python loops over a 14×14 grid for 25
ticks × population × generations.  Default config (POP=30, GENS=40)
takes ~30 s in CPython 3.12.  For larger sweeps, prefer the C
library; the Python port exists so the algorithm is portable to
any environment that runs Python — Jupyter notebooks, micropython
sketches, sandboxed REPLs.

Usage::

    from hexhunter import hexhunter, hexhunter_refine, fitness

    out = hexhunter()                     # all defaults
    out = hexhunter(population=8, generations=4, rng_seed=7)

    refined = hexhunter_refine(out, generations=10)
    print(fitness(refined))
"""

from __future__ import annotations
from dataclasses import dataclass
from typing import Callable, Optional

K              = 4
NSIT           = 16384
GENOME_BYTES   = 4096

DEF_POP                = 30
DEF_GENS               = 40
DEF_INIT_MUT_RATE      = 0.05
DEF_BREED_MUT_RATE     = 0.005
DEF_GRID_W             = 14
DEF_GRID_H             = 14
DEF_HORIZON            = 25
DEF_RNG_SEED           = 42

DY  = (-1, -1,  0,  0,  1,  1)
DXE = ( 0,  1, -1,  1, -1,  0)
DXO = (-1,  0, -1,  1,  0,  1)


# ── RNG (matches the C `hh_rng_t` / `lcg_state` exactly) ──────────────

class _Rng:
    """Park-Miller-style LCG bit-for-bit compatible with hexhunter.c."""
    __slots__ = ('state',)

    def __init__(self, seed: int):
        self.state = (seed if seed != 0 else 1) & 0xFFFFFFFF

    def u32(self) -> int:
        self.state = (self.state * 1103515245 + 12345) & 0xFFFFFFFF
        return self.state >> 16

    def rand(self) -> int:
        """Same shape as the C `hh_rand`: 0..0xFFFF."""
        return self.u32() & 0xFFFF

    def rand_unit(self) -> float:
        return self.rand() / 0xFFFF


_RAND_MAX = 0xFFFF


# ── Packed-genome helpers (2 bits per situation, K=4) ────────────────

def g_get(g: bytearray, idx: int) -> int:
    return (g[idx >> 2] >> ((idx & 3) * 2)) & 3

def g_set(g: bytearray, idx: int, v: int) -> None:
    b = idx >> 2
    o = (idx & 3) * 2
    g[b] = (g[b] & ~(3 << o)) | ((v & 3) << o)

def _sit_idx(self_: int, n: list[int]) -> int:
    i = self_
    for k in range(6):
        i = i * K + n[k]
    return i


# ── Hex stepping ─────────────────────────────────────────────────────

def _step_grid(g: bytearray, src: bytearray, dst: bytearray,
               gw: int, gh: int) -> None:
    n = [0] * 6
    for y in range(gh):
        dx = DXO if (y & 1) else DXE
        for x in range(gw):
            self_ = src[y * gw + x]
            for k in range(6):
                yy = y + DY[k]
                xx = x + dx[k]
                if 0 <= yy < gh and 0 <= xx < gw:
                    n[k] = src[yy * gw + xx]
                else:
                    n[k] = 0
            dst[y * gw + x] = g_get(g, _sit_idx(self_, n))


def _seed_grid(grid: bytearray, gw: int, gh: int, seed: int) -> None:
    r = _Rng(seed)
    for i in range(gw * gh):
        grid[i] = r.u32() & 3


# ── Class-4 fitness ──────────────────────────────────────────────────

def _fitness(genome: bytearray, grid_seed: int,
             gw: int, gh: int, horizon: int) -> tuple[float, float]:
    n = gw * gh
    a = bytearray(n)
    b = bytearray(n)
    _seed_grid(a, gw, gh, grid_seed)
    act = [0.0] * horizon

    for t in range(horizon):
        _step_grid(genome, a, b, gw, gh)
        changed = sum(1 for i in range(n) if a[i] != b[i])
        act[t] = changed / n
        a[:] = b

    uniform = all(a[i] == a[0] for i in range(1, n))
    counts = [0] * K
    for i in range(n):
        counts[a[i]] += 1
    diversity = sum(1 for c in range(K) if counts[c] * 100 >= n)

    tail_n = max(1, horizon // 3)
    avg = sum(act[horizon - tail_n:]) / tail_n

    score = 0.0
    if not uniform:
        score += 1.0
    aperiodic = any(act[i] > 0.001 for i in range(horizon - tail_n, horizon))
    if aperiodic:
        score += 1.5

    if avg <= 0.12:
        activity_reward = avg / 0.12
    else:
        activity_reward = (0.75 - avg) / 0.63
    if activity_reward < 0:
        activity_reward = 0
    score += 2.0 * activity_reward

    if diversity >= 2:
        d = min(diversity, K)
        score += 0.25 * d

    return score, avg


# ── GA ops ───────────────────────────────────────────────────────────

def _mutate(dst: bytearray, src: bytearray, rate: float, r: _Rng) -> None:
    dst[:] = src
    for i in range(NSIT):
        if r.rand_unit() < rate:
            g_set(dst, i, r.rand() & 3)


def _cross(dst: bytearray, a: bytearray, b: bytearray, r: _Rng) -> None:
    cut = 1 + (r.rand() % (GENOME_BYTES - 1))
    dst[:cut] = a[:cut]
    dst[cut:] = b[cut:]


def identity_genome() -> bytearray:
    """Every situation → self colour (1024-byte blocks of 0x00, 0x55,
    0xAA, 0xFF for self ∈ {0,1,2,3})."""
    g = bytearray(GENOME_BYTES)
    g[      :1024] = b'\x00' * 1024
    g[1024  :2048] = b'\x55' * 1024
    g[2048  :3072] = b'\xAA' * 1024
    g[3072  :4096] = b'\xFF' * 1024
    return g


# ── Public API ───────────────────────────────────────────────────────

@dataclass
class HHConfig:
    population:           int   = DEF_POP
    generations:          int   = DEF_GENS
    init_mutation_rate:   float = DEF_INIT_MUT_RATE
    breed_mutation_rate:  float = DEF_BREED_MUT_RATE
    grid_w:               int   = DEF_GRID_W
    grid_h:               int   = DEF_GRID_H
    horizon:              int   = DEF_HORIZON
    rng_seed:             int   = DEF_RNG_SEED
    progress: Optional[Callable[[int, int, float, float, float], None]] = None


def _resolve(**kwargs) -> HHConfig:
    return HHConfig(**kwargs)


def _run_ga(cfg: HHConfig, seed_genome: bytes) -> bytes:
    pop, gens = cfg.population, cfg.generations
    if pop < 2 or gens < 1:
        raise ValueError('population must be >= 2 and generations >= 1')
    gw, gh, horizon = cfg.grid_w, cfg.grid_h, cfg.horizon
    if gw < 3 or gh < 3 or horizon < 3:
        raise ValueError('grid_w/grid_h/horizon must each be >= 3')

    r = _Rng(cfg.rng_seed)
    pool: list[bytearray] = [bytearray(seed_genome)]
    for _ in range(pop - 1):
        child = bytearray(GENOME_BYTES)
        _mutate(child, pool[0], cfg.init_mutation_rate, r)
        pool.append(child)

    fits = [0.0] * pop
    last_tail = 0.0

    def _score_all() -> None:
        nonlocal last_tail
        for i in range(pop):
            s, t = _fitness(pool[i], cfg.rng_seed, gw, gh, horizon)
            fits[i] = s
            last_tail = t   # tail of the last-scored genome; for hunter
                            # parity we care about pool[0]'s tail after
                            # the sort — handled below.

    def _sort_desc() -> None:
        # Insertion sort to match C memmove pattern (stable in practice).
        for i in range(1, pop):
            fv = fits[i]
            tmp = bytearray(pool[i])
            j = i - 1
            while j >= 0 and fits[j] < fv:
                fits[j + 1] = fits[j]
                pool[j + 1] = bytearray(pool[j])
                j -= 1
            fits[j + 1] = fv
            pool[j + 1] = tmp

    for gen in range(gens):
        _score_all()
        _sort_desc()
        if cfg.progress:
            mean_s = sum(fits) / pop
            best_tail = _fitness(pool[0], cfg.rng_seed, gw, gh, horizon)[1]
            cfg.progress(gen + 1, gens, fits[0], mean_s, best_tail)

        # Breed bottom half from top half.
        for i in range(pop // 2, pop):
            pa = r.rand() % (pop // 2)
            pb = r.rand() % (pop // 2)
            tmp = bytearray(GENOME_BYTES)
            _cross(tmp, pool[pa], pool[pb], r)
            child = bytearray(GENOME_BYTES)
            _mutate(child, tmp, cfg.breed_mutation_rate, r)
            pool[i] = child

    _score_all()
    _sort_desc()
    return bytes(pool[0])


def hexhunter(*, population: int = DEF_POP,
              generations: int = DEF_GENS,
              init_mutation_rate: float = DEF_INIT_MUT_RATE,
              breed_mutation_rate: float = DEF_BREED_MUT_RATE,
              grid_w: int = DEF_GRID_W,
              grid_h: int = DEF_GRID_H,
              horizon: int = DEF_HORIZON,
              rng_seed: int = DEF_RNG_SEED,
              progress: Optional[Callable] = None) -> bytes:
    """Run the GA from scratch.  All arguments are keyword-only and
    default to the original hunter.c constants — call ``hexhunter()``
    with no arguments for the unchanged original run.  Returns the
    best 4096-byte ruleset."""
    cfg = HHConfig(population=population, generations=generations,
                    init_mutation_rate=init_mutation_rate,
                    breed_mutation_rate=breed_mutation_rate,
                    grid_w=grid_w, grid_h=grid_h, horizon=horizon,
                    rng_seed=rng_seed, progress=progress)
    return _run_ga(cfg, identity_genome())


def hexhunter_refine(in_genome: bytes, *,
                     population: int = DEF_POP,
                     generations: int = DEF_GENS,
                     init_mutation_rate: float = DEF_INIT_MUT_RATE,
                     breed_mutation_rate: float = DEF_BREED_MUT_RATE,
                     grid_w: int = DEF_GRID_W,
                     grid_h: int = DEF_GRID_H,
                     horizon: int = DEF_HORIZON,
                     rng_seed: int = DEF_RNG_SEED,
                     progress: Optional[Callable] = None) -> bytes:
    """Continue the search around an existing 4096-byte ruleset.
    Same kwargs as ``hexhunter()``.  Returns the new best 4096-byte
    ruleset."""
    if len(in_genome) != GENOME_BYTES:
        raise ValueError(f'in_genome must be {GENOME_BYTES} bytes, '
                          f'got {len(in_genome)}')
    cfg = HHConfig(population=population, generations=generations,
                    init_mutation_rate=init_mutation_rate,
                    breed_mutation_rate=breed_mutation_rate,
                    grid_w=grid_w, grid_h=grid_h, horizon=horizon,
                    rng_seed=rng_seed, progress=progress)
    return _run_ga(cfg, bytes(in_genome))


def fitness(genome: bytes, *,
            grid_w: int = DEF_GRID_W, grid_h: int = DEF_GRID_H,
            horizon: int = DEF_HORIZON,
            rng_seed: int = DEF_RNG_SEED) -> float:
    """Score a single 4096-byte ruleset under the same regime."""
    if len(genome) != GENOME_BYTES:
        raise ValueError(f'genome must be {GENOME_BYTES} bytes, '
                          f'got {len(genome)}')
    s, _ = _fitness(bytearray(genome), rng_seed, grid_w, grid_h, horizon)
    return s


# ── Self-test / CLI ──────────────────────────────────────────────────

def _cli() -> int:
    """Tiny CLI mirroring hh_cli — for verification against the C build.
    Usage: python hexhunter.py [POP] [GENS] [SEED] [OUT_PATH] [IN_PATH]"""
    import sys
    args = sys.argv[1:]
    pop  = int(args[0]) if len(args) > 0 else DEF_POP
    gens = int(args[1]) if len(args) > 1 else DEF_GENS
    seed = int(args[2]) if len(args) > 2 else DEF_RNG_SEED
    out_path = args[3] if len(args) > 3 else 'winner_py.bin'
    in_path  = args[4] if len(args) > 4 else None

    def progress(gen, total, best, mean, tail):
        print(f'  gen {gen:2d}/{total}  best={best:.3f}  mean={mean:.3f}  '
              f'best_activity_tail={tail:.3f}', file=sys.stderr)

    if in_path:
        with open(in_path, 'rb') as f:
            in_genome = f.read()
        print(f'refining {in_path} ...', file=sys.stderr)
        out = hexhunter_refine(in_genome, population=pop, generations=gens,
                                rng_seed=seed, progress=progress)
    else:
        print('evolving from identity ...', file=sys.stderr)
        out = hexhunter(population=pop, generations=gens,
                         rng_seed=seed, progress=progress)

    with open(out_path, 'wb') as f:
        f.write(out)
    final = fitness(out, rng_seed=seed)
    print(f'wrote {out_path} ({len(out)} bytes)  fitness={final:.3f}',
          file=sys.stderr)
    return 0


if __name__ == '__main__':
    raise SystemExit(_cli())
